From 22acab67c1db417976789cc7f473d63e7a606934 Mon Sep 17 00:00:00 2001
From: "Zhang, Yifan" <yifan.zhang@intel.com>
Date: Mon, 19 Nov 2018 22:53:56 +0800
Subject: [PATCH 12/25] Fix VideoView parsing some URIs failure issue

Benchmark TabletMark2017 throws errors when playback video in
Android P. Its root cause is that vold set a obb file path
including a ":", which can't be parsed correctly in VideoView
logic. This patch is to fix this issue.

Test: Run TabletMark2017.
Bug: 117532232
Change-Id: Ia12329f303f04365e7b7f9b74448f35b185783d9
Signed-off-by: Zhang, Yifan <yifan.zhang@intel.com>
Signed-off-by: Pranav Vashi <neobuddy89@gmail.com>
Signed-off-by: DennySPB <dennyspb@gmail.com>
---
 core/java/android/widget/VideoView.java   | 14 ++++++++++++++
 media/java/android/media/MediaPlayer.java | 16 +++++++++++++++-
 2 files changed, 29 insertions(+), 1 deletion(-)

diff --git a/core/java/android/widget/VideoView.java b/core/java/android/widget/VideoView.java
index 58a2b0f00877..f24235988607 100644
--- a/core/java/android/widget/VideoView.java
+++ b/core/java/android/widget/VideoView.java
@@ -48,6 +48,7 @@ import android.view.SurfaceView;
 import android.view.View;
 import android.widget.MediaController.MediaPlayerControl;
 
+import java.io.File;
 import java.io.IOException;
 import java.io.InputStream;
 import java.util.Map;
@@ -245,6 +246,19 @@ public class VideoView extends SurfaceView
      * @param path the path of the video.
      */
     public void setVideoPath(String path) {
+        /**
+         * Argument path is a valid POSIX path,
+         * but may not be a valid URI, e.g
+         * /mnt/obb/obb:0/VideoPlaybackWL/
+         * And setVideoURI requires its parameter
+         * to be a URI, thus convert local path to
+         * to URI here to avoid further uri parsing
+         * error.
+         */
+        File file = new File(path);
+        if (file.exists()) {
+            path = file.toURI().toString();
+        }
         setVideoURI(Uri.parse(path));
     }
 
diff --git a/media/java/android/media/MediaPlayer.java b/media/java/android/media/MediaPlayer.java
index d1c94c92ddef..f9f34ee1e2fd 100644
--- a/media/java/android/media/MediaPlayer.java
+++ b/media/java/android/media/MediaPlayer.java
@@ -1058,7 +1058,21 @@ public class MediaPlayer extends PlayerBase
         final String scheme = uri.getScheme();
         final String authority = ContentProvider.getAuthorityWithoutUserId(uri.getAuthority());
         if (ContentResolver.SCHEME_FILE.equals(scheme)) {
-            setDataSource(uri.getPath());
+            /**
+            * POSIX file path may contain some characters which are ambiguous for a URI,
+            * e.g. /mnt/obb/obb:0/VideoPlaybackWL/
+            * it will break further logics in setDataSource,
+            * so we have to handle the local file here.
+            */
+            final File file = new File(uri.getPath());
+            if (file.exists()) {
+                FileInputStream is = new FileInputStream(file);
+                FileDescriptor fd = is.getFD();
+                setDataSource(fd);
+                is.close();
+            } else {
+                throw new IOException("setDataSource failed.");
+            }
             return;
         } else if (ContentResolver.SCHEME_CONTENT.equals(scheme)
                 && Settings.AUTHORITY.equals(authority)) {
-- 
2.17.1

